File: app/store.py
from __future__ import annotations

import asyncio
import json
from datetime import UTC, datetime
from pathlib import Path
from typing import Any
from uuid import uuid4


class ConversationStore:
    def __init__(self, data_file: Path) -> None:
        self.data_file = data_file
        self._lock = asyncio.Lock()

    def _read(self) -> list[dict[str, Any]]:
        if not self.data_file.exists():
            return []
        return json.loads(self.data_file.read_text(encoding="utf-8"))

    def _write(self, items: list[dict[str, Any]]) -> None:
        self.data_file.parent.mkdir(parents=True, exist_ok=True)
        temporary = self.data_file.with_suffix(".tmp")
        temporary.write_text(json.dumps(items, ensure_ascii=False, indent=2), encoding="utf-8")
        temporary.replace(self.data_file)

    async def list(self) -> list[dict[str, Any]]:
        async with self._lock:
            items = self._read()
        summaries = [
            {
                "id": item["id"],
                "title": item["title"],
                "createdAt": item["createdAt"],
                "updatedAt": item["updatedAt"],
                "messageCount": len(item["messages"]),
            }
            for item in items
        ]
        return sorted(summaries, key=lambda item: item["updatedAt"], reverse=True)

    async def get(self, conversation_id: str) -> dict[str, Any] | None:
        async with self._lock:
            return next((item for item in self._read() if item["id"] == conversation_id), None)

    async def create(self) -> dict[str, Any]:
        async with self._lock:
            items = self._read()
            now = datetime.now(UTC).isoformat()
            conversation = {
                "id": str(uuid4()),
                "title": "新任务",
                "createdAt": now,
                "updatedAt": now,
                "messages": [],
            }
            items.append(conversation)
            self._write(items)
            return conversation

    async def add_message(self, conversation_id: str, role: str, content: str) -> dict[str, Any]:
        async with self._lock:
            items = self._read()
            conversation = next((item for item in items if item["id"] == conversation_id), None)
            if conversation is None:
                raise KeyError("会话不存在")
            conversation["messages"].append(
                {
                    "id": str(uuid4()),
                    "role": role,
                    "content": content,
                    "createdAt": datetime.now(UTC).isoformat(),
                }
            )
            if role == "user" and sum(m["role"] == "user" for m in conversation["messages"]) == 1:
                conversation["title"] = " ".join(content.split())[:28] or "新任务"
            conversation["updatedAt"] = datetime.now(UTC).isoformat()
            self._write(items)
            return conversation

    async def delete(self, conversation_id: str) -> bool:
        async with self._lock:
            items = self._read()
            remaining = [item for item in items if item["id"] != conversation_id]
            if len(remaining) == len(items):
                return False
            self._write(remaining)
            return True

File: src/store.js
import { mkdir, readFile, writeFile } from "node:fs/promises";
import path from "node:path";
import crypto from "node:crypto";

const dataDir = path.resolve("data");
const dataFile = path.join(dataDir, "conversations.json");

async function load() {
  try {
    return JSON.parse(await readFile(dataFile, "utf8"));
  } catch (error) {
    if (error.code === "ENOENT") return [];
    throw error;
  }
}

async function save(conversations) {
  await mkdir(dataDir, { recursive: true });
  await writeFile(dataFile, JSON.stringify(conversations, null, 2), "utf8");
}

export async function listConversations() {
  const items = await load();
  return items
    .map(({ id, title, createdAt, updatedAt, messages }) => ({
      id, title, createdAt, updatedAt, messageCount: messages.length,
    }))
    .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
}

export async function getConversation(id) {
  return (await load()).find((item) => item.id === id) ?? null;
}

export async function createConversation() {
  const items = await load();
  const now = new Date().toISOString();
  const conversation = {
    id: crypto.randomUUID(),
    title: "新任务",
    createdAt: now,
    updatedAt: now,
    messages: [],
  };
  items.push(conversation);
  await save(items);
  return conversation;
}

export async function addMessage(id, message) {
  const items = await load();
  const conversation = items.find((item) => item.id === id);
  if (!conversation) throw new Error("会话不存在");

  conversation.messages.push({
    id: crypto.randomUUID(),
    createdAt: new Date().toISOString(),
    ...message,
  });
  if (message.role === "user" && conversation.messages.filter((m) => m.role === "user").length === 1) {
    conversation.title = message.content.replace(/\s+/g, " ").slice(0, 28) || "新任务";
  }
  conversation.updatedAt = new Date().toISOString();
  await save(items);
  return conversation;
}

export async function deleteConversation(id) {
  const items = await load();
  const next = items.filter((item) => item.id !== id);
  if (next.length === items.length) return false;
  await save(next);
  return true;
}
